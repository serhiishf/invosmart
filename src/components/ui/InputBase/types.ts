import { InputHTMLAttributes } from 'react';

type InputPadding = 'default' | 'moderate' | 'minimal' | 'none';

export interface InputBaseProps extends InputHTMLAttributes<HTMLInputElement> {
  paddingLeft?: InputPadding;
  paddingRight?: InputPadding;
  isError?: boolean;
  /**
   * The isReadOnlyMode prop is used instead readonly attribute to keep inputs focusable and part of the tab order on iOS Safari, where the readonly attribute would otherwise remove them.
   */
  isReadOnlyMode?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}
