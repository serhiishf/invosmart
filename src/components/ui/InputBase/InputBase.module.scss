$_inputBasePaddingPrimary: 24px;
$_inputBasePaddingCompact: 16px;
$_inputBasePaddingMinimal: 8px;

.inputBase {
  width: 100%;
  height: 100%;
  outline: none;
  box-shadow: none;
  font-family: $inputBaseFontFamily;
  font-size: $inputBaseFontSize;
  font-weight: 500;
  color: $inputBaseTextColor;
  background-color: transparent;
  border: none;

  &--readOnlyMode {
    caret-color: transparent;
    cursor: default;
  }

  &::placeholder {
    color: $inputBasePlaceholderColor;
  }

  &:focus {
    border: none;
    outline: none;
    box-shadow: none;
  }

  &--paddingLeft {
    &-none {
      padding-left: 0;
    }
    &-default {
      padding-left: $_inputBasePaddingPrimary;
    }
    &-moderate {
      padding-left: $_inputBasePaddingCompact;
    }
    &-minimal {
      padding-left: $_inputBasePaddingMinimal;
    }
  }

  &--paddingRight {
    &-none {
      padding-right: 0;
    }
    &-default {
      padding-right: $_inputBasePaddingPrimary;
    }
    &-moderate {
      padding-right: $_inputBasePaddingCompact;
    }
    &-minimal {
      padding-right: $_inputBasePaddingMinimal;
    }
  }
}
