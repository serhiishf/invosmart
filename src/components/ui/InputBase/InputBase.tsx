import classNames from 'classnames';
import { forwardRef, useState, useEffect } from 'react';
import styles from './InputBase.module.scss';
import { InputBaseProps } from './types';
import { PaddingInput } from 'constants/theme';

const InputBase = (props: InputBaseProps, ref: React.Ref<HTMLInputElement>) => {
  const {
    paddingLeft = PaddingInput.Default,
    paddingRight = PaddingInput.Default,
    isReadOnlyMode,
    value = '',
    onChange,
    ...rest
  } = props;

  const [inputValue, setInputValue] = useState(value);

  useEffect(() => {
    setInputValue(value);
  }, [value]);

  const handleChangeIfAllowed = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (!onChange || isReadOnlyMode) return;
    setInputValue(event.target.value);
    onChange(event);
  };

  return (
    <input
      className={classNames(
        styles.inputBase,
        styles[`inputBase--paddingLeft-${paddingLeft}`],
        styles[`inputBase--paddingRight-${paddingRight}`],
        isReadOnlyMode && styles['inputBase--readOnlyMode']
      )}
      onChange={handleChangeIfAllowed}
      ref={ref}
      value={inputValue}
      {...rest}
    />
  );
};

export default forwardRef(InputBase);
