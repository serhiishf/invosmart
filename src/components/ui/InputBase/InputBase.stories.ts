import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import InputBase from './InputBase';

const meta = {
  title: 'components/ui/InputBase',
  component: InputBase,
};

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {
  args: {
    placeholder: 'Placeholder',
    value: 'Value',
    onChange: fn(),
    isReadOnlyMode: false,
  },
};
