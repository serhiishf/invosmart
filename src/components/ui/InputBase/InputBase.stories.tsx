import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import InputBase from './InputBase';

const meta = {
  title: 'components/UI/InputBase',
  component: InputBase,
  decorators: [
    (Story) => (
      <div style={{ border: '1px dashed pink' }}>
        <Story />
      </div>
    ),
  ],
  tags: ['autodocs'],
  args: {
    onChange: fn(),
  },
} satisfies Meta<typeof InputBase>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {};

export const InitialValue: Story = {
  args: {
    placeholder: 'Placeholder',
    value: 'Value',
  },
};

export const Placeholder: Story = {
  args: {
    placeholder: 'Placeholder',
  },
};

export const ReadonlyMode: Story = {
  args: {
    value: 'Readonly mode',
    isReadOnlyMode: true,
  },
};

export const Disabled: Story = {
  args: {
    value: 'Disabled',
    disabled: true,
  },
};
