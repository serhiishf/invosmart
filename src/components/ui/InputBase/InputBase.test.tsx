import { render, screen, fireEvent } from '@testing-library/react';
import InputBase from './InputBase';

describe('InputBase', () => {
  describe('Default Initialization', () => {
    it('should render correctly with default props', () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      expect(input).toBeInTheDocument();
      expect(input).not.toHaveValue();
    });

    it('should match the snapshot with default props', () => {
      const { asFragment } = render(<InputBase />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should not have a placeholder attribute', () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      expect(input).not.toHaveAttribute('placeholder');
    });

    it('should update the value on user input', () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      fireEvent.change(input, { target: { value: 'typed text' } });
    });
  });

  describe('initialization with props', () => {
    it('should render empty input with placeholder', () => {
      render(<InputBase placeholder="Some placeholder" />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveAttribute('placeholder', 'Some placeholder');
    });

    it('should render the input correctly with an initial value', () => {
      render(<InputBase value="initial value" />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveValue('initial value');
    });

    it('should not allow typing in an empty input when isReadOnlyMode is enabled', () => {
      render(<InputBase isReadOnlyMode={true} />);
      const input = screen.getByRole('textbox');
      fireEvent.change(input, { target: { value: 'typed text' } });
      expect(input).toHaveValue('');
    });

    it('should not allow typing in an input with an initial value when isReadOnlyMode is enabled', () => {
      render(<InputBase isReadOnlyMode={true} value="some text" />);
      const input = screen.getByRole('textbox');
      fireEvent.change(input, { target: { value: 'typed text' } });
      expect(input).toHaveValue('some text');
    });

    it('should match the snapshot when isReadOnlyMode is enabled', () => {
      const { asFragment } = render(<InputBase isReadOnlyMode={true} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
