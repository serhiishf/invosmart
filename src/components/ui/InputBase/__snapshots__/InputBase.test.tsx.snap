// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputBase > Default Initialization > should matches the snapshot with default props 1`] = `
<DocumentFragment>
  <input
    class="_inputBase_6ca9ba _inputBase--paddingLeft-primary_6ca9ba _inputBase--paddingRight-primary_6ca9ba"
  />
</DocumentFragment>
`;
